gossipsub v1.1: introduce message signing policy (#294)
libp2p/specs: pubsub/README.md

and
[here](https://ethresear.ch/t/improving-the-ux-of-rent-with-a-sleeping-waking-mechanism/1480)).

## Message Identification

> To uniquely identify a message in a set of topics, a `message_id` is computed based on the message.
> This can be configured on the application layer, as `message_id_fn(*Message) => message_id`.
> A `message_id_fn` may conditionally call different `message_id_fn` implementations per topic (or group thereof).

Reviewer: ```suggestion
To uniquely identify a message in a set of topics (for de-duplication, tracking, scoring and other purposes), a `message_id` is calculated based on the message.
How the calculated happens can be configured on the application layer by supplying a function `message_id_fn`, such that `message_id_fn(*Message) => message_id`.

> [[ Implementation note ]]: At the time of writing this section, go-libp2p-pubsub (reference implementation of this spec) only allows configuring a single top-level `message_id_fn`. This function may, however, vary its behaviour based on the topic (contained inside its `*Message`) argument. Thus, it's feasible to implement a per-topic policy using branch selection control flow logic. go-libp2p-pubsub plans to push down the configuration of the `message_id_fn` to the topic level. Other implementations are encouraged to do the same.
```

> The message ID approach generally fits in two flavors:

Reviewer: ```suggestion
The message ID calculation approach generally fits in two flavors:
```

- **origin-stamped** messaging: the combination of the `seqno` and `from` fields
  uniquely identifies a message based on the *author*.

> - **content-stamped** messaging: a message ID derived from the `data` field

Reviewer: ```suggestion
- **content-addressed** messaging: a message ID derived from the `data` field
```

  uniquely identifies a message based on the *data*.

> The default `message_id_fn` is origin-stamped, and defined as the string concatenation of `from` and `seqno`.

Reviewer: ```suggestion
**The default `message_id_fn` is origin-stamped,** and defined as the string concatenation of `from` and `seqno`.
```

> If fabricated collisions are not a concern, or difficult enough within the window the message is relevant in,
> a `message_id` based on a short digest of inputs may benefit performance.

Reviewer: ```suggestion
If fabricated collisions are not a concern, or difficult enough within the window the message is relevant in,
a `message_id` based on a short digest of inputs may benefit performance. Whichever the choice, it is crucial that **all peers** participating in a topic implement the same message ID calculation logic, or the topic may function suboptimally.
```

> Note that different specialized pubsub components, such as the 'timecache' used in the Go implementation,
> may use the `message_id` to key messages.

Reviewer: ```suggestion
Note that different specialized pubsub components, such as the 'timecache' used in the Go implementation, scoring functions or circuit-breakers 
may use the `message_id` to key and track messages.
```

It was also proposed in [#116](https://github.com/libp2p/specs/issues/116)
to use a `message_hash`, however, it was noted:
> a potential caveat with using hashes instead of seqnos:
the peer won't be able to send identical messages (e.g. keepalives) within the
timecache interval, as they will get rejected as duplicates.

> Some applications may not need keepalives, or choose to implement something more specific than a message hash. 

Reviewer: ```suggestion
Some applications may not need keepalives, or choose to implement something more specific than a message hash. In those cases where duplicate payloads are not desirable, a `content-based` message ID function may be more appropriate.
```

Author: I think you mean `content-addressed` here, to keep it consistent with other changes.

- **origin-stamped** messaging: the combination of the `seqno` and `from` fields
  uniquely identifies a message based on the *author*.
- **content-addressed** messaging: a message ID derived from the `data` field
  uniquely identifies a message based on the *data*.

**The default `message_id_fn` is origin-stamped,** and defined as the string concatenation of `from` and `seqno`.

If fabricated collisions are not a concern, or difficult enough within the window the message is relevant in,
a `message_id` based on a short digest of inputs may benefit performance. Whichever the choice, it is crucial that **all peers** participating in a topic implement the same message ID calculation logic, or the topic may function suboptimally.

Note that different specialized pubsub components, such as the 'timecache' used in the Go implementation, scoring functions or circuit-breakers 
may use the `message_id` to key and track messages.

It was also proposed in [#116](https://github.com/libp2p/specs/issues/116)
to use a `message_hash`, however, it was noted:
> a potential caveat with using hashes instead of seqnos:
the peer won't be able to send identical messages (e.g. keepalives) within the

> timecache interval, as they will get rejected as duplicates.

Reviewer 2: ignored, not rejected.

Messages can be optionally signed, and it is up to the peer whether to accept and forward
unsigned messages.

> When the receiver expects unsigned content-based messages, and thus does not expect

Reviewer: Mind elaborating what does `content-based messages` means here?

Reviewer: I think the two paragraphs that follow are better expressed as a bullet. Why don't we introduce the Lax modes here too? We anyway restrict all of this to v1.1, so we might as well introduce all options here.

Author: Should have called it `content-addressed`, or maybe you can think of better wording, to describe messages that identify themselves without the author, seqnoi and signature approach, like in eth2. These messages do not need to be signed, as for these it depends on the application layer, which checks contents etc.

the `from`, `seqno`, `signature`, or `key` fields, it may reject the messages (`StrictNoSign`).
And if not, the receiver may choose to enforce signatures strictly (`StrictSign`).
This optionality is configurable with the signing policy options starting from `v1.1`. 

For signing purposes, the `signature` and `key` fields are used:
- The `signature` field contains the signature.
- The `key` field contains the signing key when it cannot be inlined in the source peer ID.
  When present, it must match the peer ID.

The signature is computed over the marshalled message protobuf _excluding_ the key field.

> This includes any fields that are not recognized, but still included in the marshalled data.

Reviewer 3: this is problematic because protobuf is non-deterministic in its marshalling  - the way to write this specification would be to remove mentions of protobuf and instead explicitly outline the marshalling in terms of key-value pairs and field types - it would have to be done carefully such that the result is readable by a compliant protobuf parser - it might also happen by chance that some protobuf encoders produce a valid byte stream, but with unrecognised fields in particular, the likelihood is low.

Reviewer: Let's hoist the protobuf deterministic serialisation discussion to a new issue.

Reviewer: It's been recurrent enough that it needs to be a first-class discussion with a solid outcome.

Reviewer 2: I think we should specify canonical encoding here to remove all ambiguity.
